### PR TITLE
Composer 2 support

### PR DESCRIPTION
**Pull request by mbaynton**
Basic Composer 2 support.

I haven't exhaustively reviewed the entire plugin yet, maybe there's more work to be done. But this looks like a good start at least.

Corresponding issue #447

### CODE REVIEW

File: src/Composer/Installers/CakePHPInstaller.php [hunk 1]
```diff
@@ -58,25 +58,15 @@ protected function matchesCakeVersion($matcher, $version)
...
-            $pool = new Pool('dev');
-            $pool->addRepository($repos);
-            $packages = $pool->whatProvides('cakephp/cakephp');
-            foreach ($packages as $package) {
-                $installed = new $constraintClass('=', $package->getVersion());
-                if ($cake3->matches($installed)) {
-                    return true;
-                }
-            }
+        if (! $repositoryManager) {
```
Comment (mbaynton) [inline src/Composer/Installers/CakePHPInstaller.php]: So all this looks to be asking if `cakephp/cakephp` is >= 3, but in a complicated way that touched a lot of Composer's APIs and breaks against Composer 2. This way to answer the same question is compatible with `^1.0|^2.0`

File: tests/Composer/Installers/Test/CakePHPInstallerTest.php [hunk 2]
```diff
@@ -63,9 +63,20 @@ public function testInflectPackageVars()
     public function testGetLocations() {
         $package = new RootPackage('CamelCased', '1.0', '1.0');
         $composer = $this->composer;
+
+        $io = $this->getMock('Composer\IO\IOInterface');
+        $config = $this->getMock('Composer\Config');
+
+        // Simultaneous support for Composer 1 and 2
```
Comment (mbaynton) [inline tests/Composer/Installers/Test/CakePHPInstallerTest.php]: The test also directly touches Composer APIs that changed in 2.0; use Reflection to do the right thing

File: tests/Composer/Installers/Test/CakePHPInstallerTest.php [hunk 3]
```diff
@@ -83,11 +94,6 @@ public function testGetLocations() {
         $result = $installer->getLocations();
         $this->assertContains('Plugin/', $result['plugin']);
 
-        // special handling for 2.x versions when 3.x is still in development
```
Comment (mbaynton) [inline tests/Composer/Installers/Test/CakePHPInstallerTest.php]: This assertion will fail if left in, but I'm pretty sure nobody cares anymore as we are on CakePHP 4. Maybe should find a CakePHP contact to run it by?

File: composer.json [hunk 4]
```diff
@@ -96,11 +96,13 @@
         "roundcube/plugin-installer": "*"
     },
     "require": {
-        "composer-plugin-api": "^1.0"
+        "composer-plugin-api": "^1.0|^2.0"
     },
     "require-dev": {
-        "composer/composer": "1.0.*@dev",
-        "phpunit/phpunit": "^4.8.36"
+        "composer/composer": "1.0.0|2.0.*@dev",
```
Comment (mbaynton) [inline composer.json]: This + prefer latest/`--prefer-lowest` allows us to test the plugin against Composer 1 and Composer 2

File: composer.json [hunk 5]
```diff
@@ -96,11 +96,13 @@
...
+        "phpunit/phpunit": "^4.8.36",
+        "sebastian/comparator": ">= 1.2.4",
+        "symfony/process": ">= 2.3"
```
Comment (mbaynton) [inline composer.json]: These dependencies of dependencies have constraints that are in fact too low, so fix to minimum versions with required features.

File: composer.json [hunk 6]
```diff
@@ -96,11 +96,13 @@
         "roundcube/plugin-installer": "*"
     },
     "require": {
-        "composer-plugin-api": "^1.0"
+        "composer-plugin-api": "^1.0|^2.0"
```
Comment (Seldaek) [inline composer.json]: Please prefer `^1.0 || ^2.0` for clarity and as that's the recommended way. Same below at line 102.

**Comment by Seldaek** [comment]
Seems fine to me, but I would refrain on merging it for now or at least refrain from tagging any release if it gets merged, until 2.0 hits beta at least.

File: tests/Composer/Installers/Test/CakePHPInstallerTest.php (hunk 3 above)
Comment (jadb) [reply tests/Composer/Installers/Test/CakePHPInstallerTest.php]: I believe it's fine. @ADmad can confirm.

Comment (ADmad) [reply tests/Composer/Installers/Test/CakePHPInstallerTest.php]: I don't remember what this special handling was for. CakePHP 4 might be the latest major version but there are still many apps using CakePHP 2.x in production.

**Comment by Seldaek** [comment]
Thanks

**Pull request merged**